File: 1448-count-good-nodes-in-binary-tree/1448-count-good-nodes-in-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int goodNodes(TreeNode* root) {
        int res = 0;
        dfs(root, res, -1e5);
        return res;
    }
    
    void dfs(TreeNode *node, int &res, int prevMax){
        if(!node) return;
        
        if(node->val >= prevMax) res++;
        prevMax = max(node->val, prevMax);
        dfs(node->left, res, prevMax);
        dfs(node->right, res, prevMax);
    }
};

File: 1448-count-good-nodes-in-binary-tree/1448-count-good-nodes-in-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        res = [0]
        
        def dfs(root, prevMaxVal):
            if root==None:
                return
            
            if root.val >= prevMaxVal:
                prevMaxVal = max(root.val, prevMaxVal)
                res[0] += 1
            
            dfs(root.left, prevMaxVal)
            dfs(root.right, prevMaxVal)
        
        dfs(root, int(-1e5))
        return res[0]